### PR TITLE
Simplify dynamic mappings updates.

### PR DESCRIPTION
**Pull request by jpountz**
While dynamic mappings updates are using the same code path as updates from the
API when applied on a data node since #10593, they were still using a different
code path on the master node. This commit makes dynamic updates processed the
same way as updates from the API, which also seems to do a better way at
acknowledgements (I could not reproduce the ConcurrentDynamicTemplateTests
failure anymore). It also adds more checks, like for instance that indexing on
replicas should not trigger dynamic mapping updates since they should have been
handled on the primary before.

### CODE REVIEW

File: src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 1]
```diff
@@ -392,20 +375,52 @@ private WriteResult shardIndexOperation(BulkShardRequest request, IndexRequest i
...
+            if (update != null) {
+                final String indexName = indexService.index().name();
+                if (indexName.equals(RiverIndexName.Conf.indexName(settings))) {
+                    // With rivers, we have a chicken and egg problem if indexing
+                    // the _meta document triggers a mapping update. Because we would
+                    // like to validate the mapping update first, but on the other
+                    // hand putting the mapping would start the river, which expects
+                    // to find a _meta document
+                    // So we have no choice but to index first and send mappings afterwards
+                    indexService.mapperService().merge(indexRequest.type(), new CompressedString(update.toBytes()), true);
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: can we assigne `indexService.mapperService()` to a var? just to remove the chaining :)

File: src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java [hunk 2]
```diff
@@ -134,179 +118,9 @@ public void updateMappingOnMasterSynchronously(String index, String indexUUID, S
...
-        public final CompressedString mappingSource;
-        public final MappingUpdateListener listener;
-
-        MappingChange(String index, String indexUUID, String type, CompressedString mappingSource, MappingUpdateListener listener) {
-            this.index = index;
-            this.indexUUID = indexUUID;
-            this.type = type;
-            this.mappingSource = mappingSource;
-            this.listener = listener;
+    public void updateMappingOnMasterSynchronously(String index, String type, Mapping mappingUpdate, TimeValue timeout) throws Throwable {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java]: I wonder where the uuid went?

**Comment by s1monw** [comment]
left a couple of comments

File: src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java (hunk 2 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java]: It used to be necessary but now the update is essentially about calling putMapping on the IndicesAdminClient which does not need the index uuid

File: src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 1 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Ok will do

File: src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java (hunk 2 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java]: so I wonder if we should pass it on though. I think it's good to validate it and if we use the client as a user of ES we just pass `IndexMetaData#INDEX_UUID_NA_VALUE` as the default?

**Comment by s1monw** [comment]
LGTM I think we should add the UUID in a second step in a followup issue

**Pull request merged**